kernel/app_io/src/lib.rs

```diff
@@ -1,329 +1,226 @@
+        self.lock().read(buf)
     }
 }
 
-mod shared_maps {
-    use spin::{Mutex, MutexGuard};
-    use alloc::collections::BTreeMap;
-    use IoControlFlags;
-    use IoStreams;
-
-    /// Map a task to its IoControlFlags structure.
-    /// When shells call `insert_child_streams`, the default value is automatically inserted for that task.
-    /// When shells call `remove_child_streams`, the corresponding structure is removed from this map.
-    static APP_IO_CTRL_FLAGS: Mutex<BTreeMap<usize, IoControlFlags>> = Mutex::new(BTreeMap::new());
-
-    /// Map a task id to its IoStreams structure.
-    /// Shells should call `insert_child_streams` when spawning a new app,
-    /// which effectively stores a new key value pair to this map. 
-    /// After a shell's child app exits, the shell should call `remove_child_streams` to clean it up.
-    static APP_IO_STREAMS: Mutex<BTreeMap<usize, IoStreams>> = Mutex::new(BTreeMap::new());
-
-    /// Lock and returns the `MutexGuard` of `APP_IO_CTRL_FLAGS`. Use `lock_all_maps()` if
-    /// you want to lock both of the maps to avoid deadlock.
-    pub fn lock_flag_map() -> MutexGuard<'static, BTreeMap<usize, IoControlFlags>> {
-        APP_IO_CTRL_FLAGS.lock()
+impl ImmutableWrite for StdioWriter {
+    fn write(&self, buf: &[u8]) -> io::Result<usize> {
+        self.lock().write(buf)
     }
+}
 
-    /// Lock and returns the `MutexGuard` of `APP_IO_STREAMS`. Use `lock_all_maps()` if
-    /// you want to lock both of the maps to avoid deadlock.
-    pub fn lock_stream_map() -> MutexGuard<'static, BTreeMap<usize, IoStreams>> {
-        APP_IO_STREAMS.lock()
+impl ImmutableRead for Slave {
+    fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
+        self.read(buf)
     }
+}
 
-    /// Lock two maps `APP_IO_CTRL_FLAGS` and `APP_IO_STREAMS` at the same time and returns a
-    /// tuple containing their `MutexGuard`s. This function exerts a sequence of locking when
-    /// we need to lock more than one of them. This prevents deadlock.
-    pub fn lock_all_maps() -> (MutexGuard<'static, BTreeMap<usize, IoControlFlags>>,
-                               MutexGuard<'static, BTreeMap<usize, IoStreams>>) {
-        (APP_IO_CTRL_FLAGS.lock(), APP_IO_STREAMS.lock())
+impl ImmutableWrite for Slave {
+    fn write(&self, buf: &[u8]) -> io::Result<usize> {
+        self.write(buf)
     }
 }
 
-
-/// An application can call this function to get the terminal to which it should print.
-pub fn get_my_terminal() -> Option<Arc<Mutex<Terminal>>> {
-    task::get_my_current_task_id()
-        .and_then(|id| shared_maps::lock_stream_map()
-            .get(&id)
-            .map(|property| property.terminal.clone())
-        )
+/// Stores the stdio queues, key event queue and the pointer to the terminal
+/// for applications. This structure is provided for application's use and only
+/// contains necessary one-end readers/writers to queues. On the shell side, we
+/// have full control to queues.
```

Comment:
"full control to queues"? small typo?

Reply:
I didn't write that but I updated it to be correct.

Reply:
oh my apologies. Thanks for fixing it anyway.